### PR TITLE
[Experimental] Reuse upstream vLLM models on PyTorch MPS

### PR DESCRIPTION
**Pull request by lucamotz**
The current Metal runner loads model implementations through MLX, so supporting an upstream vLLM architecture still depends on a separate MLX implementation and adapter. This draft explores a smaller compatibility boundary: execute **upstream vLLM model classes and weight loaders on PyTorch MPS**, while reusing upstream request handling and sampling.

`VLLM_METAL_MODEL_BACKEND=torch` selects an experimental out-of-tree platform. The existing MLX path remains the default. The implementation adds four backend files (about 720 lines including types, comments, and whitespace), with no changes to vLLM core and no copied model implementations.

### Reuse and scope

- `get_model` / `ModelRegistry` own architecture resolution and weight loading.
- `CPUModelRunner` owns request updates, batching, slot mapping, KV allocation/binding, prompt logprobs, and sampling. A small wrapper transfers inputs to MPS and logits to CPU; parameters, hidden states, and KV storage stay on MPS.
- Attention consumes `CommonAttentionMetadata` and uses native PyTorch SDPA over vLLM's paged cache. It supports causal/chunked prefill, decode, GQA, prefix reuse, and sliding windows. The cache adapter accepts both the 0.28 split layout and main's packed per-layer layout without copying cache storage.
- Setup rejects features outside the initial dense, unquantized, single-device, eager text-generation scope. An explicit but unavailable MPS selection fails after plugin probing, where vLLM cannot silently swallow the error and fall back to CPU.
- KV storage defaults to at most 512 MiB, bounded by system/MPS headroom. An explicit `kv_cache_memory_bytes` is checked against that budget. This is a KV limit, not a total-memory guarantee.

This is **not full upstream model coverage or a performance claim**. MoE, SSM/hybrid models, quantization, multimodal/pooling, LoRA, speculative decoding, distributed execution, and transfer/offload paths need further backend work. CUDA-only operations in otherwise dense architectures may still fail. SDPA gathers visible cache tensors and expands GQA heads, so long-context scratch memory and throughput need work. The normal macOS vLLM CPU extension is required for the inherited slot-mapping operation; an `empty` build is insufficient.

The alternative loader-only approach would still have to translate upstream Torch forward contexts and KV caches into the MLX runner's model-specific contract. Keeping model execution behind the existing upstream runner avoids that additional model-porting layer. Details and usage are in `docs/pytorch_backend.md`.

### Validation

Tested on Apple M4, 24 GB unified RAM, macOS 26.6.2, Python 3.12.13, PyTorch 2.13.0, the official vLLM 0.28 macOS CPU wheel, and this repository's pinned MLX dependencies. Large-model and performance suites were not run given the machine's memory pressure.

```bash
python -m pytest tests/test_register.py tests/test_macos_defaults.py \
  tests/test_platform.py tests/test_pytorch_attention.py \
  tests/test_pytorch_model_runner.py tests/test_tensor_bridge.py -q
# 137 passed, 10 skipped

python -m pytest tests/test_pytorch_models.py -q
# 4 passed

ruff check .
ruff format --check .
mypy vllm_metal
# All pass; 299 files formatted, 141 source files type-checked.
```

The model tests create tiny Llama, Qwen2, and Qwen3 checkpoints locally. Greedy token IDs match both Hugging Face CPU and `mlx_lm` references, including chunked prefill and repeated prefix-cache use. They assert that actual upstream model classes and all parameters are on MPS, and exercise logit bias plus prompt/output logprobs. Attention tests include real MPS float32/float16/bfloat16 execution, physical block addressing, released sliding-window prefix blocks, and both cache layouts. Memory-budget tests allocate no large buffers.

Additional manual checks:

- **SmolLM2-135M**, revision `93efa2f097d58c2a74874c7e644dbc9b0cee75a2`: all **36/36 greedy generated token IDs** matched the HF CPU reference across three prompts, in float32 with a **16 MiB KV cache**. Post-generation MPS allocation was 557,486,848 bytes; this is not a peak-process-memory measurement.
- A normally spawned localhost `vllm serve` process returned HTTP 200 for `/health` and `/v1/completions`, with expected continuation text and output logprobs. SIGTERM cleanup logged a forced-worker shutdown/semaphore warning; clean server shutdown needs further validation.
- Against upstream vLLM Python sources at **`537af2c3a4ba7462ddc9bc94ec7a4ea496da6d2e`**, tiny Llama/Qwen2/Qwen3 greedy checks and Llama chunked-prefill/prefix-cache/sampling-control checks passed. These used the installed 0.28 macOS wheel's C extension through a source overlay, **not a fresh native build of main**. The plugin's vLLM version pin is unchanged.

### Draft review

Checked open vllm-metal and upstream vLLM proposals on 2026-09-07. No open PR implements this upstream-model/MPS runner path. Existing MLX sampling, model-adapter, and KV-offload proposals operate on the existing MLX path; this does not replace their work or the version-migration checklist in #700.

AI assistance was used for investigation, implementation, and tests. This is intentionally a draft for human review of the reuse boundary before expanding model coverage or optimizing kernels. Broader checkpoint/dtype evaluation and a fresh current-main macOS build remain follow-up validation.

### CODE REVIEW

**Review by WindChimeRan — COMMENTED** [review]
Quick feedback:

I also don't like relying on `mlx_lm` / `mlx_vlm` for model configs, because they tend to be pretty messy. For example, `mlx_lm` always trims the MTP head, and support for visual encoders mostly depends on `mlx_vlm`, which is maintained by a smaller community.

But switching to vLLM models is non-trivial because vLLM is deeply coupled with PyTorch. From my own benchmarks, PyTorch MPS is far slower than MLX, which makes it basically unusable on Apple Silicon.

So I'd still vote for maintaining only the MLX runtime backend, even if that means keeping a number of ugly patches. I'm not sure whether it's feasible to read vLLM's model configs and translate them cleanly into MLX, though...

**Review by LxYuan0420 — COMMENTED** [review]
Thanks for exploring this, but I’m not sold on this direction for vllm-metal. A separate PyTorch MPS runtime would bypass our optimised Metal kernels, support fewer features, and significantly increase our maintenance burden :)

**Comment by lucamotz** [comment]
@WindChimeRan @LxYuan0420 thanks for the quick feedback haha! Didn’t expect you guys to even look at a draft. I’ll just close it again then, taking it on board. We would have run some benchmarks at work over the next few weeks to see if we can get this to work, but if you say it’s truly not worth it, we’ll rely on other software for our Mac experiments. It’s a shame really!
